internal/provider/resource_application.go

```diff
@@ -875,6 +1086,61 @@ func (r *applicationResource) Update(ctx context.Context, req resource.UpdateReq
 		updateApplicationInput.EndpointBindings = endpointBindings
 	}
 
+	// Check if we have new storages in plan that not existed in the state, and add their constraints to the
+	// update application input
+	if !plan.Storage.Equal(state.Storage) {
+		var planStorageSlice, stateStorageSlice []nestedStorage
+		resp.Diagnostics.Append(plan.Storage.ElementsAs(ctx, &planStorageSlice, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		resp.Diagnostics.Append(state.Storage.ElementsAs(ctx, &stateStorageSlice, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// Map of storage name to storage constraints
+		planStorageMap := make(map[string]jujustorage.Constraints)
+		for _, storage := range planStorageSlice {
+			// Validate storage size
+			parsedSize, err := utils.ParseSize(storage.Size.ValueString())
+			if err != nil {
+				return
```

Comment:
Before returning you need to call resp.Diagnostics.AddError() or this won't be caught by terraform.